crates/builder/src/tasks/tx_poller.rs

```diff
@@ -0,0 +1,119 @@
+use std::time::Duration;
+use std::{collections::HashMap, time};
+
+use alloy_consensus::TxEnvelope;
+use alloy_primitives::TxHash;
+use eyre::Error;
+use reqwest::Client;
+use serde_json::from_str;
+use serde_json::Value;
+use tokio::sync::mpsc;
+use tokio::task::JoinHandle;
+
+pub use crate::config::BuilderConfig;
+
+/// Implements a poller for the block builder to pull transactions from the transaction pool.
+pub struct TxPoller {
+    // config for the builder
+    pub config: BuilderConfig,
+    // Reqwest client for fetching transactions from the tx-pool
+    pub client: Client,
+    //  Maintain a set of transaction hashes to their expiration times
+    pub seen_txns: HashMap<TxHash, time::Instant>,
+}
+
+/// TxPoller implements a poller that fetches unique transactions from the transaction pool.
+impl TxPoller {
+    /// returns a new TxPoller with the given config.
+    pub fn new(config: &BuilderConfig) -> Self {
+        Self {
+            config: config.clone(),
+            client: Client::new(),
+            seen_txns: HashMap::new(),
+        }
+    }
+
+    /// polls the tx-pool for unique transactions and evicts expired transactions.
+    /// unique transactions that haven't been seen before are sent into the builder pipeline.
+    pub async fn check_tx_pool(&mut self) -> Result<Vec<TxEnvelope>, Error> {
+        let mut unique: Vec<TxEnvelope> = Vec::new();
+        let result = self.client.get(self.config.tx_pool_url.to_string() + "/get").send().await?;
+
+        // parse the response as a JSON array of key/value pairs of transaction hashes to transaction envelopes
+        let parsed: Value = from_str(&result.text().await?)?;
+        if let Value::Array(items) = parsed.clone() {
+            for item in items {
+                if let Value::Object(map) = item {
+                    if let Some(Value::String(value)) = map.get("value") {
+                        if let Ok(parsed) = from_str::<TxEnvelope>(value) {
+                            self.check_cache(parsed, &mut unique);
+                        }
+                    }
+                }
+            }
+        }
+
+        Ok(unique)
+    }
+
+    /// checks if the transaction has been seen before and if not, adds it to the unique transactions list.
+    fn check_cache(&mut self, tx: TxEnvelope, unique: &mut Vec<TxEnvelope>) {
+        self.seen_txns.entry(*tx.tx_hash()).or_insert_with(|| {
+            // add to unique transactions
+            unique.push(tx.clone());
+            // expiry is now + cache_duration
+            time::Instant::now() + Duration::from_secs(self.config.tx_pool_cache_duration)
+        });
+    }
+
+    /// removes entries from seen_txns that have lived past expiry
+    fn evict(&mut self) {
+        let expired_keys: Vec<TxHash> = self.seen_txns
+            .iter()
+            .filter_map(|(key, &expiration)| {
+                if expiration.elapsed() > Duration::from_secs(0) {
+                    Some(key.clone())
```

Comment:
you should be able to do `*key` instead of `key.clone()` as `key` is a `Copy` type